### ISSUE TITLE
heap-buffer-overflow read in wtf_parse1() and wtf_parse()

### ISSUE THREAD

**Issue by kcwu** [tats/w3m#68]
input (`xxd cases/tats-w3m-68.1`)
```
00000000: 3c6d 6574 6120 6368 6172 7365 743d 6762  <meta charset=gb
00000010: 3138 3033 303e 0a3c 696d 6720 7372 633d  18030>.<img src=
00000020: 803e                                     .>
```

how to reproduce:
```
ASAN_OPTIONS=abort_on_error=1:detect_leaks=0 LD_LIBRARY_PATH=./notgc ./w3m-tats.asan -T text/html -dump cases/tats-w3m-68.1
```

stderr:
```
=================================================================
==2732258==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200000c4d2 at pc 0x0000007585fb bp 0x7ffeb14ed730 sp 0x7ffeb14ed728
READ of size 1 at 0x60200000c4d2 thread T0
    #0 0x7585fa in wtf_parse1 /targets/w3m-tats/libwc/wtf.c:405:16
    #1 0x759ff3 in wtf_parse /targets/w3m-tats/libwc/wtf.c:473:7
    #2 0x6fcbd6 in wc_conv_to_ces /targets/w3m-tats/libwc/conv.c:93:34
    #3 0x6fc040 in wc_Str_conv /targets/w3m-tats/libwc/conv.c:23:9
    #4 0x6fd953 in wc_Str_conv_strict /targets/w3m-tats/libwc/conv.c:37:10
    #5 0x691b08 in url_encode /targets/w3m-tats/url.c:2296:12
    #6 0x569c7e in process_img /targets/w3m-tats/file.c:3239:9
    #7 0x586bc6 in HTMLtagproc1 /targets/w3m-tats/file.c:4920:8
    #8 0x5927e3 in HTMLlineproc0 /targets/w3m-tats/file.c:6477:10
    #9 0x5a5a0c in loadHTMLstream /targets/w3m-tats/file.c:7252:2
    #10 0x55bbf8 in loadHTMLBuffer /targets/w3m-tats/file.c:6781:5
    #11 0x55ee64 in loadSomething /targets/w3m-tats/file.c:224:16
    #12 0x5535ac in loadGeneralFile /targets/w3m-tats/file.c:2241:6
    #13 0x4f9202 in main /targets/w3m-tats/main.c:1017:12
    #14 0x7f187dd6df44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287
    #15 0x41bf25 in _start (/w3m-tats.asan+0x41bf25)

0x60200000c4d2 is located 0 bytes to the right of 2-byte region [0x60200000c4d0,0x60200000c4d2)
allocated by thread T0 here:
    #0 0x4c6288 in __interceptor_malloc (/w3m-tats.asan+0x4c6288)
    #1 0x7f187f4bbc21 in GC_malloc_atomic /notgc/notgc.c:275
    #2 0x691a8b in url_encode /targets/w3m-tats/url.c:2296:12
    #3 0x569c7e in process_img /targets/w3m-tats/file.c:3239:9
    #4 0x586bc6 in HTMLtagproc1 /targets/w3m-tats/file.c:4920:8
    #5 0x5927e3 in HTMLlineproc0 /targets/w3m-tats/file.c:6477:10
    #6 0x5a5a0c in loadHTMLstream /targets/w3m-tats/file.c:7252:2
    #7 0x55bbf8 in loadHTMLBuffer /targets/w3m-tats/file.c:6781:5
    #8 0x55ee64 in loadSomething /targets/w3m-tats/file.c:224:16
    #9 0x5535ac in loadGeneralFile /targets/w3m-tats/file.c:2241:6
    #10 0x4f9202 in main /targets/w3m-tats/main.c:1017:12
    #11 0x7f187dd6df44 in __libc_start_main /build/eglibc-oGUzwX/eglibc-2.19/csu/libc-start.c:287

SUMMARY: AddressSanitizer: heap-buffer-overflow /targets/w3m-tats/libwc/wtf.c:405:16 in wtf_parse1
Shadow bytes around the buggy address:
  0x0c047fff9840: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff9850: fa fa fa fa fa fa fa fa fa fa 00 fa fa fa 02 fa
  0x0c047fff9860: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 06 fa
  0x0c047fff9870: fa fa fd fa fa fa 00 00 fa fa 03 fa fa fa fd fa
  0x0c047fff9880: fa fa 00 00 fa fa 00 00 fa fa 05 fa fa fa fd fa
=>0x0c047fff9890: fa fa fd fa fa fa 00 00 fa fa[02]fa fa fa 00 00
  0x0c047fff98a0: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 00 02
  0x0c047fff98b0: fa fa 00 00 fa fa 00 06 fa fa fd fd fa fa 00 00
  0x0c047fff98c0: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 04 fa
  0x0c047fff98d0: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 00 01
  0x0c047fff98e0: fa fa 00 00 fa fa 00 00 fa fa 00 00 fa fa 04 fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==2732258==ABORTING
```

This is detected with help of dummy libgc wrapper. See http://github.com/kcwu/fuzzing-w3m/notgc for detail.
More detail to reproduce please see http://github.com/kcwu/fuzzing-w3m

For your convenience,
gdbline:
LD_LIBRARY_PATH=./notgc ASAN_OPTIONS=abort_on_error=1:detect_leaks=0 gdb --args ./w3m-tats.asan -T text/html -dump cases/tats-w3m-68.1

----
There are many variant cases with similar stack. I feel they are the same issue. For example,

input (`xxd cases/tats-w3m-68.2`)
```
00000000: 3c6d 6574 6120 6368 6172 7365 743d 6762  <meta charset=gb
00000010: 3138 3033 303e 0a3c 6120 7461 7267 6574  18030>.<a target
00000020: 3d80 3e                                  =.>
```
```
    #0 0x7585fa in wtf_parse1 /targets/w3m-tats/libwc/wtf.c:405:16
    #1 0x759ff3 in wtf_parse /targets/w3m-tats/libwc/wtf.c:473:7
    #2 0x6fd4ec in wc_conv_to_ces /targets/w3m-tats/libwc/conv.c:101:31
    #3 0x6fc040 in wc_Str_conv /targets/w3m-tats/libwc/conv.c:23:9
    #4 0x6fd953 in wc_Str_conv_strict /targets/w3m-tats/libwc/conv.c:37:10
    #5 0x59a7e4 in HTMLlineproc2body /targets/w3m-tats/file.c:5687:8
```

input (`xxd cases/tats-w3m-68.3`)
```
00000000: 3c74 6162 6c65 3e30 3c62 7220 3c3e 303c  <table>0<br <>0<
00000010: 786d 703e 3c68 723e 3030 3030 3030 3030  xmp><hr>00000000
00000020: 303c 696e 7465 726e 616c 3e30 3030 3030  0<internal>00000
00000030: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
00000040: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
00000050: 3030 3030 3030 3030 3030 3030 3030 3030  0000000000000000
00000060: 3030 3030 2030 3030 3030 3030 3030 3030  0000 00000000000
00000070: 3030 3c69 6e70 7574 3e                   00<input>
```
```
    #0 0x7585fa in wtf_parse1 /targets/w3m-tats/libwc/wtf.c:405:16
    #1 0x75bba8 in wtf_get_code /targets/w3m-tats/libwc/wtf.c:534:11
    #2 0x5b45ae in conv_symbol /targets/w3m-tats/file.c:7605:22
```

input (`xxd cases/tats-w3m-68.4`)
```
00000000: 3c74 6162 6c65 3e30 3c6e 6f62 7220 3c3e  <table>0<nobr <>
00000010: 303c 786d 703e c8ab 3c64 6976 3e3c 696e  0<xmp>..<div><in
00000020: 7465 726e 616c 3e3c 696e 7075 745f 616c  ternal><input_al
00000030: 7420 6669 643d 303e 3c66 6f72 6d3e 3030  t fid=0><form>00
00000040: 30                                       0
```
```
    #0 0x75a1d8 in wtf_parse /targets/w3m-tats/libwc/wtf.c:474:53
    #1 0x6fcbd6 in wc_conv_to_ces /targets/w3m-tats/libwc/conv.c:93:34
```

This is found by afl-fuzz.

**Comment by tats** [comment]
Fixed, thank you.

**Comment by kcwu** [comment]
cases/tats-w3m-68.3 is not fixed yet.

**Comment by kcwu** [comment]
cases/tats-w3m-68.3 seems different issue.

**Comment by kcwu** [comment]
cases/tats-w3m-68.3: in form_update_line(),
```
323         buf = New_N(char, len);
```
I don't fully understand the logic of `form_update_line()`. IIUC, `buf` is allocated with size `len` and filled `len` bytes of data. `form_update_line()` forgot to preserve one byte for end of string character. Because no end of string, `wtf_parse1()` segfault when reach end of the buffer later.

**Comment by tats** [comment]
> cases/tats-w3m-68.3: in form_update_line(),

Fixed, thank you.
cf. https://github.com/tats/w3m/issues/82

**Comment by tats** [comment]
Fixed again for cases/tats-w3m-68.3.

**Comment by tats** [comment]
Patch updated due to https://github.com/tats/w3m/issues/86
